### PR TITLE
Add custom commands & options to Rush

### PR DESCRIPTION
**Pull request by nickpape-msft**

### CODE REVIEW

**Comment by nickpape-msft** [comment]
@pgonzal @iclanton @qz2017 Can you please take a look at this implementation? I will not merge it until after Ian merges the ClickToRun change!

File: apps/rush-lib/src/data/CommandLineConfiguration.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    if (commandLineJson) {
+      if (commandLineJson.customCommands) {
+        commandLineJson.customCommands.forEach((command: ICustomCommand) => {
+          this.commands.set(command.name, command);
+        });
+      }
+
+      if (commandLineJson.customOptions) {
+        Object.keys(commandLineJson.customOptions).forEach((flagName: string) => {
+          const customOption: ICustomOption = commandLineJson.customOptions![flagName];
```
Comment (qz2017) [inline apps/rush-lib/src/data/CommandLineConfiguration.ts]: This one could be undefined. '!' does not help.

File: apps/rush/src/actions/CustomRushAction.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+              return enumValue.name;
+            })
+        });
+      }
+    });
+  }
+
+  private _collectTelemetry(stopwatch: Stopwatch, success: boolean): void {
+    const extraData: { [key: string]: string } = {
+      to: (!!this._toFlag.value).toString(),
```
Comment (qz2017) [inline apps/rush/src/actions/CustomRushAction.ts]: I am thinking of prefixing the extraData fields to become "build_to", "build_from", and the new" build + customOption.longName". Can you include that change in this pullrequest?

File: apps/rush/src/utilities/TaskManager.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  RushConfiguration,
+  RushConfigurationProject,
+  IRushLinkJson
+} from '@microsoft/rush-lib';
+import { JsonFile } from '@microsoft/node-core-library';
+
+import TaskRunner from '../taskRunner/TaskRunner';
+import ProjectBuildTask from '../taskRunner/ProjectBuildTask';
+
+export class TaskManager {
```
Comment (qz2017) [inline apps/rush/src/utilities/TaskManager.ts]: Can you add tests for this class?

File: apps/rush/src/actions/CustomRushAction.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+interface ICustomOptionInstance extends ICustomOption {
+  longName: string;
+  parameterValue: CommandLineFlagParameter | CommandLineOptionParameter;
+}
+
+export class CustomRushAction extends BaseRushAction {
+  private customOptions: Array<ICustomOptionInstance> = [];
+
+  private _fromFlag: CommandLineStringListParameter;
```
Comment (qz2017) [inline apps/rush/src/actions/CustomRushAction.ts]: Do all customRushActions support "from" and "to" by default? That is cool.

Comment (nickpape-msft) [reply apps/rush/src/actions/CustomRushAction.ts]: Yep!

---
In reply to: [145582119](https://github.com/Microsoft/web-build-tools/pull/387#discussion_r145582119) [](ancestors = 145582119)

File: apps/rush/src/utilities/TaskManager.ts (hunk 3 above)
Comment (nickpape-msft) [reply apps/rush/src/utilities/TaskManager.ts]: Yeah, why not

---
In reply to: [145581577](https://github.com/Microsoft/web-build-tools/pull/387#discussion_r145581577) [](ancestors = 145581577)

File: apps/rush-lib/src/cli/actions/BuildAction.ts [hunk 5]
```diff
@@ -1,43 +0,0 @@
...
-    this._cleanParameter = this.defineFlagParameter({
-      parameterLongName: '--clean',
-      parameterShortName: '-c',
-      description: 'Skip incremental build detection and force a clean build. Same as the "rebuild" command.'
-    });
-  }
-
-  protected onExecute(): void {
-    // If the clean flag is false, we will support incremental build by default.
-    this._isIncrementalBuildAllowed = !this._cleanParameter.value;
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/actions/BuildAction.ts]: >_isIncrementalBuildAllowed [](start = 9, length = 26)

double check this value is being looked at

File: apps/rush-lib/src/cli/actions/CustomCommandFactory.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    }));
+
+    customActions.set('rebuild', new CustomRushAction(parser, {
+      actionVerb: 'rebuild',
+      summary: 'Clean and rebuild the entire set of projects',
+      documentation: documentationForBuild
+    }));
+
+    commandLineConfig.commands.forEach((command: ICustomCommand) => {
+      customActions.set(command.name, new CustomRushAction(parser, {
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/actions/CustomCommandFactory.ts]: >command [](start = 24, length = 7)

throw if it is already set

File: apps/rush-lib/src/cli/actions/CustomCommandFactory.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        summary: command.description,
+        documentation: command.description
+      }));
+    });
+
+    commandLineConfig.options.forEach((customOption: ICustomOption, longName: string) => {
+      customOption.supportedCommands.forEach((associatedCommand: string) => {
+        const customAction: CustomRushAction | undefined = customActions.get(associatedCommand);
+        if (customAction) {
+          customAction.addCustomOption(longName, customOption);
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/actions/CustomCommandFactory.ts]: >addCustomOption [](start = 23, length = 15)

don't let them define the same option twice

File: apps/rush-lib/src/data/commandLineConfiguration.schema.json [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+            "supportedCommands"
+          ],
+
+          "properties": {
+            "description": {
+              "title": "Description",
+              "type": "string"
+            },
+
+            "supportedCommands": {
```
Comment (pgonzal) [inline apps/rush-lib/src/data/commandLineConfiguration.schema.json]: >supportedCommands [](start = 13, length = 17)

Looking at your sample file, I think maybe "associatedCommands" would be a more intuitive name, since this is a reverse mapping

File: apps/rush-lib/src/data/CommandLineConfiguration.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+import * as path from 'path';
+
+import {
+  JsonFile,
+  JsonSchema
+} from '@microsoft/node-core-library';
+
+import { RushConstants } from '../RushConstants';
+
+/** @public */
```
Comment (pgonzal) [inline apps/rush-lib/src/data/CommandLineConfiguration.ts]: Why are these marked as @public?

File: common/reviews/api/rush-lib.api.ts [hunk 10]
```diff
@@ -92,6 +92,15 @@ enum ChangeType {
...
 }
 
+// @public
+class CommandLineConfiguration {
+  // (undocumented)
+  public commands: Map<string, ICustomCommand>;
+  // (undocumented)
+  public options: Map<string, ICustomOption>;
+  public static tryLoadFromFile(jsonFilename: string): CommandLineConfiguration;
+}
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: Don't expose this.  @iclanton is supposed to be working to trim down the API to a minimal set of configuration parsing functionality, now that we eliminated the rush/rush-lib boundary

File: libraries/ts-command-line/src/CommandLineParameterProvider.ts [hunk 11]
```diff
@@ -81,8 +81,17 @@ abstract class CommandLineParameterProvider {
   }
 
   protected defineOptionParameter(options: ICommandLineOptionDefinition): CommandLineOptionParameter {
+    if (!options.options || options.options.length < 2) {
```
Comment (pgonzal) [inline libraries/ts-command-line/src/CommandLineParameterProvider.ts]: >options.options [](start = 9, length = 15)

weird

File: libraries/ts-command-line/src/CommandLineParameterProvider.ts [hunk 12]
```diff
@@ -81,8 +81,17 @@ abstract class CommandLineParameterProvider {
...
+      throw new Error(`When defining an option parameter, the options array must`
```
Comment (pgonzal) [inline libraries/ts-command-line/src/CommandLineParameterProvider.ts]: >options [](start = 62, length = 7)

This is a confusing name. A command line "option" is usually just a string parameter, or maybe it means any general parameter.  If there are a fixed set of choices, that's usually called an "enum".

If this is a disruptive change, we can handle it in a separate PR

File: apps/rush-lib/src/cli/actions/CustomRushAction.ts [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+  CommandLineStringListParameter,
+  CommandLineOptionParameter,
+  ICommandLineActionOptions
+} from '@microsoft/ts-command-line';
+
+import RushCommandLineParser from './RushCommandLineParser';
+import { BaseRushAction } from './BaseRushAction';
+import { TaskManager } from '../utilities/TaskManager';
+
+interface ICustomOptionInstance extends ICustomOption {
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/actions/CustomRushAction.ts]: >ICustomOptionInstance [](start = 10, length = 21)

Why do we need this?

File: apps/rush-lib/src/cli/actions/CustomRushAction.ts [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+  private _parallelismParameter: CommandLineIntegerParameter;
+
+  private _parser: RushCommandLineParser;
+
+  constructor(parser: RushCommandLineParser, options: ICommandLineActionOptions) {
+    super(options);
+    this._parser = parser;
+  }
+
+  public addCustomOption(longName: string, option: ICustomOption): void {
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/actions/CustomRushAction.ts]: >public [](start = 2, length = 6)

Document this

File: apps/rush-lib/src/cli/actions/CustomRushAction.ts [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+      key: 'PROJECT2',
+      description: 'Build all projects that directly or indirectly depend on the specified project'
+    });
+    this._verboseParameter = this.defineFlagParameter({
+      parameterLongName: '--verbose',
+      parameterShortName: '-v',
+      description: 'Display the logs during the build, rather than just displaying the build status summary'
+    });
+
+    // @TODO we should throw if they are trying to overwrite built in flags
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/actions/CustomRushAction.ts]: >// @TODO  [](start = 4, length = 9)

?

File: apps/rush-lib/src/cli/actions/CustomRushAction.ts [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+          parameterShortName: customOption.shortName,
+          parameterLongName: customOption.longName,
+          description: customOption.description
+        });
+      } else if (customOption.optionType === 'enum') {
+        customOption.parameterValue = this.defineOptionParameter({
+          parameterShortName: customOption.shortName,
+          parameterLongName: customOption.longName,
+          description: customOption.description,
+          defaultValue: (customOption as ICustomOption as ICustomEnumOption).defaultValue,
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/actions/CustomRushAction.ts]: >as ICustomOption as ICustomEnumOption [](start = 38, length = 37)

?

File: apps/rush-lib/src/data/CommandLineConfiguration.ts [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+export interface ICustomEnumValue {
+  name: string;
+  description: string;
+}
+
+/** @public */
+export interface ICustomOption {
+  optionType: 'enum' | 'flag';
```
Comment (pgonzal) [inline apps/rush-lib/src/data/CommandLineConfiguration.ts]: > optionType: 'enum' | 'flag'; [](start = 1, length = 29)

Use a tagged union

File: apps/rush-lib/src/data/CommandLineConfiguration.ts [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+ */
+export class CommandLineConfiguration {
+  public options: Map<string, ICustomOption>;
+  public commands: Map<string, ICustomCommand>;
+
+  /** Attempts to load pinned versions configuration from a given file */
+  public static tryLoadFromFile(jsonFilename: string): CommandLineConfiguration {
+    let commandLineJson: ICommandLineConfigurationJson | undefined = undefined;
+    if (fs.existsSync(jsonFilename)) {
+      const schemaFilename: string = path.join(__dirname, 'commandLineConfiguration.schema.json');
```
Comment (pgonzal) [inline apps/rush-lib/src/data/CommandLineConfiguration.ts]: >commandLineConfiguration [](start = 59, length = 24)

Do we use dashes?  Check the other .schema.json files

File: apps/rush-lib/src/data/commandLineConfiguration.schema.json [hunk 19]
```diff
@@ -0,0 +1,145 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
```
Comment (pgonzal) [inline apps/rush-lib/src/data/commandLineConfiguration.schema.json]: >"$schema": "http://json-schema.org/draft-04/schema#", [](start = 2, length = 53)

Check that this validates, since JsonSchema isn't doing that currently

File: apps/rush-lib/src/data/CommandLineConfiguration.ts [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+      commandLineJson = JsonFile.loadAndValidate(jsonFilename, JsonSchema.fromFile(schemaFilename));
```
Comment (pgonzal) [inline apps/rush-lib/src/data/CommandLineConfiguration.ts]: >JsonSchema.fromFile(schemaFilename) [](start = 63, length = 35)

The convention is to put this in a static member

File: apps/rush-lib/src/data/CommandLineConfiguration.ts [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+
+    if (commandLineJson) {
+      if (commandLineJson.customCommands) {
+        commandLineJson.customCommands.forEach((command: ICustomCommand) => {
+          this.commands.set(command.name, command);
+        });
+      }
+
+      if (commandLineJson.customOptions) {
+        Object.keys(commandLineJson.customOptions).forEach((flagName: string) => {
```
Comment (pgonzal) [inline apps/rush-lib/src/data/CommandLineConfiguration.ts]: >forEach [](start = 51, length = 7)

Consider for..of instead of forEach

File: apps/rush-lib/src/data/CommandLineConfiguration.ts [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+    this.commands = new Map<string, ICustomCommand>();
+
+    if (commandLineJson) {
+      if (commandLineJson.customCommands) {
+        commandLineJson.customCommands.forEach((command: ICustomCommand) => {
+          this.commands.set(command.name, command);
+        });
+      }
+
+      if (commandLineJson.customOptions) {
```
Comment (pgonzal) [inline apps/rush-lib/src/data/CommandLineConfiguration.ts]: >if [](start = 6, length = 2)

Can we write some unit tests for these edge cases?

File: apps/rush-lib/src/cli/actions/CustomRushAction.ts [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+
+    const tasks: TaskManager = new TaskManager(
+      this._parser.rushConfig,
+      this._toFlag.value,
+      this._fromFlag.value,
+      this.options.actionVerb,
+      customFlags,
+      isQuietMode,
+      this._parallelismParameter.value,
+      this.options.actionVerb === 'build'
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/actions/CustomRushAction.ts]: If this can't be simplified, maybe at least put it in an options interface.  Otherwise it will be easy to accidentally mix up the parameters

File: apps/rush-lib/src/cli/utilities/TaskManager.ts [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+  RushConfiguration,
+  RushConfigurationProject,
+  IRushLinkJson
+} from '../../index';
+import { JsonFile } from '@microsoft/node-core-library';
+
+import TaskRunner from '../taskRunner/TaskRunner';
+import ProjectBuildTask from '../taskRunner/ProjectBuildTask';
+
+export class TaskManager {
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/utilities/TaskManager.ts]: Document this:

Why we need it?
Who calls it and when?
How is it implemented?

Comment (pgonzal) [inline apps/rush-lib/src/cli/utilities/TaskManager.ts]: >TaskManager [](start = 13, length = 11)

Find a better name:

TaskSelector?

File: apps/rush-lib/src/cli/utilities/TaskManager.ts [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+    toFlags: Array<string>,
+    fromFlags: Array<string>,
+    private _commandToRun: string,
+    private _customFlags: string[],
+    isQuietMode: boolean,
+    parallelism: number,
+    private _isIncrementalBuildAllowed: boolean) {
+
+    this._taskRunner = new TaskRunner(isQuietMode, parallelism);
+    this._rushLinkJson = JsonFile.load(this._rushConfiguration.rushLinkJsonFilename);
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/utilities/TaskManager.ts]: >rushLinkJsonFilename [](start = 63, length = 20)

Check what happens if this throws

File: apps/rush-lib/src/cli/taskRunner/ProjectBuildTask.ts [hunk 26]
```diff
@@ -15,8 +15,6 @@ import {
   RushConfiguration,
   RushConfigurationProject,
   RushConstants,
-  ErrorDetector,
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/taskRunner/ProjectBuildTask.ts]: >ErrorDetector [](start = 2, length = 13)

make sure these get deleted

File: apps/rush-lib/src/cli/taskRunner/ProjectBuildTask.ts [hunk 27]
```diff
@@ -128,23 +106,25 @@ export default class ProjectBuildTask implements ITaskDefinition {
...
+            writer.writeLine(`The clean command was registered in the package.json but is blank. Skipping 'clean' step...`);
+          } else {
+            writer.writeLine(cleanCommand);
+            try {
+              Utilities.executeShellCommand(cleanCommand, projectFolder, process.env, true);
+            } catch (error) {
+              throw new Error(`There was a problem running the 'clean' script: ${os.EOL} ${error.toString()}`);
+            }
           }
         }
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/taskRunner/ProjectBuildTask.ts]: remove

File: apps/rush-lib/src/cli/taskRunner/ProjectBuildTask.ts [hunk 28]
```diff
@@ -211,35 +176,34 @@ export default class ProjectBuildTask implements ITaskDefinition {
     }
   }
 
-  private _getBuildCommand(): string {
-    const build: string | undefined =
-      this._getScriptCommand('test') || this._getScriptCommand('build');
+  private _isBuildCommand(): boolean {
+    return this._commandToRun === 'build' || this._commandToRun === 'rebuild';
+  }
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/taskRunner/ProjectBuildTask.ts]: delete

File: apps/rush-lib/src/cli/taskRunner/ProjectBuildTask.ts [hunk 29]
```diff
@@ -211,35 +176,34 @@ export default class ProjectBuildTask implements ITaskDefinition {
...
 
-    if (build === undefined) {
-      // tslint:disable-next-line:max-line-length
-      throw new Error(`The project [${this._rushProject.packageName}] does not define a 'test' or 'build' command in the 'scripts' section of its package.json`);
-    }
+  private _getScriptToRun(): string {
+    let script: string | undefined = undefined;
+    if (this._isBuildCommand()) {
+      script = this._getScriptCommand('test') || this._getScriptCommand('build');
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/taskRunner/ProjectBuildTask.ts]: >cript = this._getScriptCommand('test') || this._getScriptCommand('build'); [](start = 7, length = 74)

remove

File: apps/rush-lib/src/cli/actions/RebuildAction.ts [hunk 30]
```diff
@@ -1,293 +0,0 @@
...
-    this._minimalParameter = this.defineFlagParameter({
-      parameterLongName: '--minimal',
-      parameterShortName: '-m',
-      description: 'Invokes the build script with the "--minimal" option, which speeds up the build by running the ' +
-        'minimal set of tasks required to produce executable output'
-    });
-    this._verboseParameter = this.defineFlagParameter({
-      parameterLongName: '--verbose',
-      parameterShortName: '-v',
-      description: 'Display the logs during the build, rather than just displaying the build status summary'
```
Comment (pgonzal) [inline apps/rush-lib/src/cli/actions/RebuildAction.ts]: If all these flags are deleted, what happens to someone who currently relies on this flags in their repo? When they upgrade to the new Rush, will the flags all be mysteriously missing?

Maybe we could include the legacy settings in an "examples" folder, so people can simply copy it to their common/config/rush folder when they migrate?

File: apps/rush-lib/src/cli/actions/CustomRushAction.ts (hunk 13 above)
Comment (nickpape-msft) [reply apps/rush-lib/src/cli/actions/CustomRushAction.ts]: It stores the parameter value

---
In reply to: [146732192](https://github.com/Microsoft/web-build-tools/pull/387#discussion_r146732192) [](ancestors = 146732192)

File: apps/rush-lib/src/data/commandLineConfiguration.schema.json (hunk 19 above)
Comment (nickpape-msft) [reply apps/rush-lib/src/data/commandLineConfiguration.schema.json]: this should be built into node-core-library

---
In reply to: [146732552](https://github.com/Microsoft/web-build-tools/pull/387#discussion_r146732552) [](ancestors = 146732552)

Comment (nickpape-msft) [reply apps/rush-lib/src/data/commandLineConfiguration.schema.json]: out of scope

---
In reply to: [147496188](https://github.com/Microsoft/web-build-tools/pull/387#discussion_r147496188) [](ancestors = 147496188,146732552)

File: apps/rush-lib/src/cli/actions/CustomRushAction.ts (hunk 15 above)
Comment (nickpape-msft) [reply apps/rush-lib/src/cli/actions/CustomRushAction.ts]: we just let ts-command-line throw if they do this.

---
In reply to: [146732263](https://github.com/Microsoft/web-build-tools/pull/387#discussion_r146732263) [](ancestors = 146732263)

File: apps/rush-lib/src/cli/actions/RebuildAction.ts (hunk 30 above)
Comment (nickpape-msft) [reply apps/rush-lib/src/cli/actions/RebuildAction.ts]: Yes, all the flags will be missing

---
In reply to: [146734888](https://github.com/Microsoft/web-build-tools/pull/387#discussion_r146734888) [](ancestors = 146734888)

File: apps/rush-lib/src/examples/Legacy_CustomCommandLine.json [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+      "supportedCommands": [ "build", "rebuild" ],
+      "shortName": "-m"
+    },
+
+    "--clean": {
+      "optionType": "flag",
+      "description": "This flag ensures that the build is done clean!",
+      "supportedCommands": [ "build", "rebuild" ],
+      "shortName": "-c"
+    }
```
Comment (nickpape-msft) [inline apps/rush-lib/src/examples/Legacy_CustomCommandLine.json]: this might  need to be removed?

File: common/reviews/api/rush-lib.api.ts [hunk 32]
```diff
@@ -286,13 +250,11 @@ class PinnedVersionsConfiguration {
...
 
 // @public
-export function RegexErrorDetector(regex: RegExp,
-    getError: (match: RegExpExecArray) => TaskError | undefined): IErrorDetectionRule;
-
-// @public
 class RushConfiguration {
   public readonly approvedPackagesPolicy: ApprovedPackagesPolicy;
   public readonly changesFolder: string;
+  // WARNING: The type "CommandLineConfiguration" needs to be exported by the package (e.g. added to index.ts)
```
Comment (pgonzal) [inline common/reviews/api/rush-lib.api.ts]: >WARNING [](start = 5, length = 7)

Fix this

**Comment by pgonzal** [comment]
> // Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.

Probably we can delete this entirely

---
Refers to: apps/rush-lib/src/gulp2vs.ts:1 in 9710fcd. [](commit_id = 9710fcd24489d2331e50be8e92fe7d21d59156ef, deletion_comment = False)

File: apps/rush-lib/src/RushConstants.ts [hunk 33]
```diff
@@ -83,6 +83,12 @@ export namespace RushConstants {
   export const packageDepsFilename: string = 'package-deps.json';
 
   /**
+   * Custom command line configuration file, which is used by rush for implementing
+   * custom command and options.
+   */
+  export const commandLineFilename: string = 'command-line.json';
```
Comment (pgonzal) [inline apps/rush-lib/src/RushConstants.ts]: >export [](start = 2, length = 6)

@iclanton you're likely to get a merge conflict

File: apps/rush-lib/src/examples/Legacy_CustomCommandLine.json [hunk 34]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (pgonzal) [inline apps/rush-lib/src/examples/Legacy_CustomCommandLine.json]: Should you add a "$schema"?

It's okay to use a URL that is a 404 now, but will work by the time we publish

Comment (pgonzal) [inline apps/rush-lib/src/examples/Legacy_CustomCommandLine.json]: Add a code comment at the top of this file saying something like this:

// This file recreates the command-line options that were built into Rush prior to version 4.
// Note that earlier versions of Rush also performed the following transformations to
// produce the command-line that was passed to the NPM scripts in the project's
// package.json file:
//
// - If "--vso" was passed to Rush, then "--color" was passed to the script instead
// - Before running a "build" script, the "clean" script was automatically executed first
//
// Rush 4 no longer performs these sorts of transformations.  Instead, it aims for
// the Rush options and script options to be as similar as possible.

Comment (pgonzal) [reply apps/rush-lib/src/examples/Legacy_CustomCommandLine.json]: (and maybe we don't need the "--color" in this example)

---
In reply to: [147541732](https://github.com/Microsoft/web-build-tools/pull/387#discussion_r147541732) [](ancestors = 147541732)

File: apps/rush-lib/src/examples/Sample_CustomCommandLine.json [hunk 35]
```diff
@@ -0,0 +1,58 @@
+{
```
Comment (pgonzal) [inline apps/rush-lib/src/examples/Sample_CustomCommandLine.json]: Add a comment saying something like:

// This example illustrates how to define your own custom command line
// options for Rush.  These options will be passed to the NPM scripts in your
// package.json file.
//
// To use this file, rename it to command-line.json and copy it into to your
// common/config/rush folder.  For more information about custom commands,
// see the documentation at [https://aka.ms/rush](https://aka.ms/rush)

**Comment by pgonzal** [comment]
I approved this PR, but FYI @iclanton is working to clean up the **rush-lib** API for this release, so please ensure that your *.api.ts changes are:

- Only exporting APIs we really want to support
- Free of any API Extractor warnings
- Documented using writing suitable for the [web site](https://microsoft.github.io/web-build-tools/api/node-core-library.html)

**Comment by qz2017** [comment]
@nickpape-msft , are you going to squash and merge or just merge? The commit messages in this pullrequest are not in a good form and will pollute the history if not squashed.

**Comment by nickpape-msft** [comment]
@qz2017 I generally prefer to squash. IMO intermediate commits are rarely useful.
